File: Core/Process.cs
using System;
using System.Collections.Generic;

namespace Core
{
    public static class Process
    {
        private static List<Handle> handles;

        internal static void RegisterHandle(Handle handle)
        {
            handles.Add(handle);
        }

        public static Optional<Error> Run()
        {
            // main event loop
            while (true)
            {
                Syscalls.EventWait();
            }
            return new Optional<Error>(Error.NotImplemented);
        }

        public static void End()
        {
            EmitDebug("Process end");
            Syscalls.ProcessEnd();
            while (true) ;
        }

        public static Optional<Error> EmitInformation(string informationText)
        {
            var oldColor = Console.ForegroundColor;
            Console.ForegroundColor = ConsoleColor.Gray;
            Console.WriteLine("Information: " + informationText);
            Console.ForegroundColor = oldColor;

            return Syscalls.ProcessEmit(Storm.SyscallProcessEmitType.Information, Error.None, informationText);
        }

        public static Optional<Error> EmitDebug(string debugText)
        {
            var oldColor = Console.ForegroundColor;
            Console.ForegroundColor = ConsoleColor.Green;
            Console.WriteLine("Debug: " + debugText);
            Console.ForegroundColor = oldColor;

            return Syscalls.ProcessEmit(Storm.SyscallProcessEmitType.Debug, Error.None, debugText);
        }

        public static Optional<Error> EmitWarning(string warningText)
        {
            var oldColor = Console.ForegroundColor;
            Console.ForegroundColor = ConsoleColor.Yellow;
            Console.WriteLine("Warning: " + warningText);
            Console.ForegroundColor = oldColor;

            return Syscalls.ProcessEmit(Storm.SyscallProcessEmitType.Warning, Error.None, warningText);
        }

        public static Optional<Error> EmitError(Error error, string errorText)
        {
            var oldColor = Console.ForegroundColor;
            Console.ForegroundColor = ConsoleColor.Red;
            Console.WriteLine("Error: " + error.ToString() + ": " + errorText);
            Console.ForegroundColor = oldColor;

            return Syscalls.ProcessEmit(Storm.SyscallProcessEmitType.Error, error, errorText);
        }
    }
}

File: rust/chaos-core/src/process.rs
use crate::error::Error;
use crate::syscalls;

pub fn emit_debug(information_text: &str) -> Option<Error> {
    syscalls::process_emit(syscalls::EmitType::Debug, Error::None, Some(information_text))
}

pub fn emit_information(information_text: &str) -> Option<Error> {
    syscalls::process_emit(syscalls::EmitType::Information, Error::None, Some(information_text))
}

pub fn emit_warning(information_text: &str) -> Option<Error> {
    syscalls::process_emit(syscalls::EmitType::Warning, Error::None, Some(information_text))
}

pub fn emit_error(error: Error, information_text: &str) -> Option<Error> {
    syscalls::process_emit(syscalls::EmitType::Error, error, Some(information_text))
}

pub fn run() -> Error {
    // this is the main event loop of an application
    while (true) {
        let result = syscalls::event_wait();
        match result {
            Ok(handle, action) => {

            },
            Err(error) => {
                return error;
            }
        }
    }

    Error::None
}

pub fn end() -> () {

}